llms/mlx_lm/models/dbrx.py

```diff
@@ -0,0 +1,268 @@
+
+        scores = mx.softmax(
+            mx.take_along_axis(gates, inds, axis=-1).astype(mx.float32),
+            axis=-1,
+        ).astype(gates.dtype)
+
+        if self.training:
+            mx.eval(inds)
+            inds = np.array(inds)
+            y = mx.zeros((x.shape[0], ne, x.shape[-1]), x.dtype)
+            for e, expert in enumerate(self.experts):
+                idx1, idx2 = map(mx.array, np.where(inds == e))
+                if idx1.size == 0:
+                    continue
+                y[idx1, idx2] = expert(x[idx1])
+
+            y = (y * scores[:, :, None]).sum(axis=1)
+        else:
+            y = []
+            for xt, st, it in zip(x, scores, inds.tolist()):
+                yt = mx.concatenate([self.experts[e](xt)[:, None] for e in it], axis=-1)
+                yt = (yt * st).sum(axis=-1)
+                y.append(yt[None, :])
+            y = mx.concatenate(y)
+
+        return y.reshape(orig_shape)
+
+
+class DecoderLayer(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.ffn = SparseMoeBlock(args)
+        self.norm_attn_norm = NormAttnNorm(args)
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Tuple[mx.array, mx.array]] = None,
+    ) -> mx.array:
+        r, h, cache = self.norm_attn_norm(x, mask, cache)
+        out = self.ffn(h) + r
+        return out, cache
+
+
+class DBRX(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.vocab_size = args.vocab_size
+        self.wte = nn.Embedding(args.vocab_size, args.d_model)
+        self.blocks = [DecoderLayer(args=args) for _ in range(args.n_layers)]
+        self.norm_f = nn.LayerNorm(args.d_model, bias=False)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        h = self.wte(inputs)
+
+        mask = None
+        T = h.shape[1]
+        if T > 1:
+            mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
+            mask = mask.astype(h.dtype)
+
+        if cache is None:
+            cache = [None] * len(self.blocks)
+
+        for e, layer in enumerate(self.blocks):
+            import pdb
+
+            h, cache[e] = layer(h, mask, cache[e])
+
+        return self.norm_f(h), cache
+
+
+class Model(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.model_type = args.model_type
+        self.transformer = DBRX(args)
+        self.lm_head = nn.Linear(args.d_model, args.vocab_size, bias=False)
+        self.args = args
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        out, cache = self.transformer(inputs, cache)
+        return self.lm_head(out), cache
+
+    @property
+    def layers(self):
+        return self.transformer.blocks
+
+    def sanitize(self, weights):
```

Comment:
The split expert can possibly be simplified with np.split(). This is simpler for splitting the tensors. Something like this, not 100% sure: 
```python
def sanitize(self, weights):
    num_experts = self.args.ffn_config["moe_num_experts"]
    dim = self.args.ffn_config["ffn_hidden_size"]

    new_weights = {k: v for k, v in weights.items() if "experts.mlp" not in k}
    expert_weights = []

    for k, v in weights.items():
        if "experts.mlp" in k:
            expert_splits = np.split(v, num_experts, axis=0)
            for e in range(num_experts):
                new_key = k.replace(".mlp.", f".{e}.weight")
                expert_weight = expert_splits[e]
                if k.endswith(".w2"):
                    expert_weight = expert_weight.T
                expert_weights.append((new_key, expert_weight))

    new_weights.update(expert_weights)

    return new_weights
```

Reply:
Nice, I like the idea of using split! I will borrow this!